src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -25,252 +26,989 @@
 #include "precompiled.hpp"
 #include "gc/shared/tlab_globals.hpp"
 #include "gc/shenandoah/shenandoahFreeSet.hpp"
+#include "gc/shenandoah/shenandoahFreeSet.inline.hpp"
 #include "gc/shenandoah/shenandoahHeap.inline.hpp"
 #include "gc/shenandoah/shenandoahHeapRegionSet.hpp"
 #include "gc/shenandoah/shenandoahMarkingContext.inline.hpp"
 #include "logging/logStream.hpp"
 #include "memory/resourceArea.hpp"
 #include "runtime/orderAccess.hpp"
 
-ShenandoahFreeSet::ShenandoahFreeSet(ShenandoahHeap* heap, size_t max_regions) :
-  _heap(heap),
-  _mutator_free_bitmap(max_regions, mtGC),
-  _collector_free_bitmap(max_regions, mtGC),
-  _max(max_regions)
+static const char* partition_name(ShenandoahFreeSetPartitionId t) {
+  switch (t) {
+    case NotFree: return "NotFree";
+    case Mutator: return "Mutator";
+    case Collector: return "Collector";
+    default: return "Unrecognized";
+  }
+}
+
+size_t ShenandoahSimpleBitMap::count_leading_ones(ssize_t start_idx) const {
```

Comment:
Are you sure that you can't use the BitMap class in src/hotspot/share/utilities/bitMap.*? Or maybe only src/hotspot/share/utilities/count_leading_zeros.hpp or src/hotspot/share/utilities/count_trailing_zeros.hpp? That would emit CPU instructions that do the counting really fast, if possible. I.e. reverse all bits after loading, then find leading/trailing zeros, problem solved real fast. I would even argue to augment the BitMap class with count_* methods as needed, using the CPU-optimized routines.

Reply:
I will see if I can exploit the services in src/hotspot/share/utilities to improve performance.

Reply:
Why use ssize_t for start_idx? or, at all? In bitMap.hpp, it declares a 'typedef size_t idx_t' and use that for bit and word indices. I'd try to stick to the same pattern as much as reasonably possible.